Avoid JIT imports in BodoSQL C++ backend

ehsantn opened the pull request:
## Changes included in this PR

<!-- Include a brief description of the changes presented in this PR and any extra context that might be helpful for reviewers. -->
Splits JIT and non-JIT code in BodoSQL to allow lazy JIT import (if necessary) in the new C++ backend.

## Testing strategy

<!-- 
Before requesting review, verify that your changes pass PR CI by adding "[run ci]" to your commit message (or add a new blank commit with that message) or explain why CI is not necessary (e.g. docs changes). 

Briefly mention how this change is tested e.g. "new unit tests added". To pass automated coverage checks, ensure that you have added `# pragma: no cover` to jitted functions. 

Ensure that newly added tests work locally on 3 ranks using both SPMD and spawn mode (default) when applicable. For example:

SPMD mode: 
  `export BODO_SPAWN_MODE=0;
  mpiexec -n 3 pytest -svW ignore bodo/tests/test_dataframe.py::my_new_test`

Spawn mode (default mode): 
  `export BODO_NUM_WORKERS=3;
  pytest -svW ignore bodo/tests/test_dataframe.py::my_new_test`
-->
Opened an issue to check for avoiding JIT imports in tests since needs some work.

## User facing changes

<!-- Mention any changes to user facing APIs here and ensure that the documentation is up to date in Bodo/docs/docs -->
None.

## Checklist
- [x] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [x] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md) 
- [x] I have installed + ran pre-commit hooks.

DrTodd13 reviewed (approved):
Might be helpful on this type of PR to note when code that is moved is also changed.  Also, might help to describe categories of changes and reasoning so you can judge at that level once rather than try to figure out the reason for 50 identical changes.  Anyway, everything looks reasonable to me.

ehsantn commented on BodoSQL/bodosql/context.py:
```diff
@@ -1595,32 +1172,226 @@ def update_schema(
+    return read_line
+
+
+def create_java_dynamic_parameter_type_list(dynamic_params_list: list[Any]):
```
Moved here with changes to avoid importing JIT for typing unnecessarily.

ehsantn commented on BodoSQL/bodosql/context_ext.py:
```diff
@@ -689,3 +696,398 @@ def overload_convert_to_pandas(
     )
+
+
+# Scalar dtypes for supported Bodo Arrays
```
All this code is just moved here.

ehsantn commented:
> Might be helpful on this type of PR to note when code that is moved is also changed. Also, might help to describe categories of changes and reasoning so you can judge at that level once rather than try to figure out the reason for 50 identical changes. Anyway, everything looks reasonable to me.

Thanks. Makes sense. Added some comments for the next reviewer.
